File: bot-api/dotnet/src/events/DisconnectedEvent.cs
using Newtonsoft.Json;

namespace Robocode.TankRoyale.BotApi.Events
{
  /// <summary>
  /// Event occurring when bot gets disconnected from server.
  /// </summary>
  public sealed class DisconnectedEvent : ConnectionEvent
  {
    /// <summary>Indication if closing of the connection was initiated by the remote host.</summary>
    public bool Remote { get; }

    /// <summary>
    /// Initializes a new instance of the DisconnectedEvent class.
    /// </summary>
    /// <param name="serverUri">URI of the server.</param>
    /// <param name="remote">Indication if closing of the connection was initiated by the remote host.</param>
    [JsonConstructor]
    public DisconnectedEvent(System.Uri serverUri, bool remote) : base(serverUri) => Remote = remote;
  }
}

File: bot-api/typescript/src/events/DisconnectedEvent.ts
import { ConnectionEvent } from "./ConnectionEvent.js";

/** Event occurring when disconnected from the server. */
export class DisconnectedEvent extends ConnectionEvent {
  readonly remote: boolean;
  readonly statusCode?: number;
  readonly reason?: string;

  constructor(serverUri: string, remote: boolean, statusCode?: number, reason?: string) {
    super(serverUri);
    this.remote = remote;
    if (statusCode !== undefined) this.statusCode = statusCode;
    if (reason !== undefined) this.reason = reason;
  }
}
